include/deal.II/base/mpi.h

```diff
@@ -1313,158 +1313,184 @@ namespace Utilities
-    {
-      return MPI_UNSIGNED_SHORT;
-    }
+        inline MPI_Datatype
+        mpi_type_id(const unsigned short *)
+        {
+          return MPI_UNSIGNED_SHORT;
+        }
 
 
 
-    inline MPI_Datatype
-    mpi_type_id(const unsigned int *)
-    {
-      return MPI_UNSIGNED;
-    }
+        inline MPI_Datatype
+        mpi_type_id(const unsigned int *)
+        {
+          return MPI_UNSIGNED;
+        }
 
 
 
-    inline MPI_Datatype
-    mpi_type_id(const unsigned long int *)
-    {
-      return MPI_UNSIGNED_LONG;
-    }
+        inline MPI_Datatype
+        mpi_type_id(const unsigned long int *)
+        {
+          return MPI_UNSIGNED_LONG;
+        }
 
 
 
-    inline MPI_Datatype
-    mpi_type_id(const unsigned long long int *)
-    {
-      return MPI_UNSIGNED_LONG_LONG;
-    }
+        inline MPI_Datatype
+        mpi_type_id(const unsigned long long int *)
+        {
+          return MPI_UNSIGNED_LONG_LONG;
+        }
 
 
 
-    inline MPI_Datatype
-    mpi_type_id(const float *)
-    {
-      return MPI_FLOAT;
-    }
+        inline MPI_Datatype
+        mpi_type_id(const float *)
+        {
+          return MPI_FLOAT;
+        }
 
 
 
-    inline MPI_Datatype
-    mpi_type_id(const double *)
-    {
-      return MPI_DOUBLE;
-    }
+        inline MPI_Datatype
+        mpi_type_id(const double *)
+        {
+          return MPI_DOUBLE;
+        }
 
 
 
-    inline MPI_Datatype
-    mpi_type_id(const long double *)
-    {
-      return MPI_LONG_DOUBLE;
-    }
+        inline MPI_Datatype
+        mpi_type_id(const long double *)
+        {
+          return MPI_LONG_DOUBLE;
+        }
 
 
 
-    inline MPI_Datatype
-    mpi_type_id(const std::complex<float> *)
-    {
-      return MPI_COMPLEX;
-    }
+        inline MPI_Datatype
+        mpi_type_id(const std::complex<float> *)
+        {
+          return MPI_COMPLEX;
+        }
 
 
 
-    inline MPI_Datatype
-    mpi_type_id(const std::complex<double> *)
-    {
-      return MPI_DOUBLE_COMPLEX;
-    }
+        inline MPI_Datatype
+        mpi_type_id(const std::complex<double> *)
+        {
+          return MPI_DOUBLE_COMPLEX;
+        }
 #  endif
+#endif
+      } // namespace MPIDataTypes
+    }   // namespace internal
+
+
+
+    /**
+     * A template variable that translates from the data type given as
+     * template argument to the corresponding
+     * `MPI_Datatype` to be used for MPI communication.
+     *
+     * As an example, the value of `mpi_type_id<int>` is `MPI_INT`. A
+     * common way to use this variable is when sending an object `obj`
+     * via MPI functions to another process, and using
+     * `mpi_type_id<decltype(obj)>` to infer the correct MPI type to
+     * use for the communication.
+     */
+    template <typename T>
+    const MPI_Datatype mpi_type_id = internal::MPIDataTypes::mpi_type_id(
+      static_cast<std::remove_cv_t<
+        std::remove_reference_t<std::remove_all_extents_t<T>>> *>(nullptr));
```

Comment:
So this is a template definition that will be implicitly instantiated creating a weak symbol to  `const MPI_Datatype mpi_type_id<T>` whenever `dealii/base/utilities.h` is included ...

Reply:
No, that's not correct. It's a *declaration* of a variable, the fact that it has an initializer notwithstanding. It's the same issue as with `static const` member variables.

In most cases, the compiler will simply replace all references to the variable by its value. You only get in trouble if someone decides to take the address of the variable (or the compiler does so if you pass this variable by-reference), which is why you need to provide the definition in the `.cc` file.

Reply:
Is there a good reason to allow array types via `std::remove_all_extents`? I would say that there is no equivalent `MPI` type for `int[3]`. If we allow it, though, I would also allow pointers, i.e., `int*` -> `MPI_INT`.

Reply:
Wolfgang, you have a bug in your example above, the first ``&buffer`` needs to be ``buffer`` because it is a pointer already. With that, the examples no longer match very well.
I am not sure if allowing pointers types as T would be more confusing then not allowing it.